src/Kestrel.Core/Internal/Http/ChunkWriter.cs

```diff
@@ -21,19 +22,20 @@ private static ArraySegment<byte> CreateAsciiByteArraySegment(string text)
 
         public static ArraySegment<byte> BeginChunkBytes(int dataCount)
         {
-            var bytes = new byte[10]
-            {
-                _hex[((dataCount >> 0x1c) & 0x0f)],
-                _hex[((dataCount >> 0x18) & 0x0f)],
-                _hex[((dataCount >> 0x14) & 0x0f)],
-                _hex[((dataCount >> 0x10) & 0x0f)],
-                _hex[((dataCount >> 0x0c) & 0x0f)],
-                _hex[((dataCount >> 0x08) & 0x0f)],
-                _hex[((dataCount >> 0x04) & 0x0f)],
-                _hex[((dataCount >> 0x00) & 0x0f)],
-                (byte)'\r',
-                (byte)'\n',
-            };
+            var bytes = new byte[10];
+            ref var r = ref bytes[0];
+            ref var hex = ref _hex[0];
+
+            Unsafe.Add(ref r, 0) = Unsafe.Add(ref hex, (dataCount >> 0x1c) & 0x0f);
```

Comment:
You should not need any of this unsafe code. This should be able to eliminate the bounds checks just fine.

Reply:
(for bytes). For `_hex` - not yet.

Reply:
Could drop the `_hex` loopkup table and use?

```csharp
// digit = (dataCount >> 0x1c) & 0x0f;
= (byte)(digit + (digit < 10 ? (byte)'0' : (byte)'a'));
```

Reply:
@jkotas thx! I got distracted by the generated `jae` and `jae SHORT`.

For `_hex` I'll try the idea from @benaadams

Reply:
``` ini

BenchmarkDotNet=v0.10.14, OS=Windows 10.0.17134
Intel Core i7-7700HQ CPU 2.80GHz (Kaby Lake), 1 CPU, 8 logical and 4 physical cores
Frequency=2742192 Hz, Resolution=364.6718 ns, Timer=TSC
.NET Core SDK=2.1.300-rc1-008673
  [Host]     : .NET Core 2.1.0-rc1 (CoreCLR 4.6.26426.02, CoreFX 4.6.26426.04), 64bit RyuJIT
  DefaultJob : .NET Core 2.1.0-rc1 (CoreCLR 4.6.26426.02, CoreFX 4.6.26426.04), 64bit RyuJIT


```
|             Method |     Mean |     Error |    StdDev | Scaled | ScaledSD |  Gen 0 | Allocated |
|------------------- |---------:|----------:|----------:|-------:|---------:|-------:|----------:|
|            Default | 17.51 ns | 0.3823 ns | 0.3755 ns |   1.00 |     0.00 | 0.0127 |      40 B |
|             Unsafe | 15.43 ns | 0.1269 ns | 0.1187 ns |   0.88 |     0.02 | 0.0127 |      40 B |
| BytesSafeHexUnsafe | 13.88 ns | 0.1718 ns | 0.1607 ns |   0.79 |     0.02 | 0.0127 |      40 B |
|           NoLookup | 14.28 ns | 0.1124 ns | 0.0938 ns |   0.82 |     0.02 | 0.0127 |      40 B |
| NoLookupBranchless | 16.28 ns | 0.3119 ns | 0.2918 ns |   0.93 |     0.03 | 0.0127 |      40 B |

`Default` is _dev_-branch
`Unsafe` this PR so far
[BytesSafeHexUnsafe](https://github.com/gfoidl/Benchmarks/blob/0fe785ac88273e977adf44c1dcd8891651316420/aspnet/Kestrel/CorrelationIdGenerator_PR2579/CorrelationIdGenerator/ChunkWriters/ChunkWriter2.cs) doesn't use `Unsafe` for `bytes`
[NoLookup](https://github.com/gfoidl/Benchmarks/blob/0fe785ac88273e977adf44c1dcd8891651316420/aspnet/Kestrel/CorrelationIdGenerator_PR2579/CorrelationIdGenerator/ChunkWriters/ChunkWriter3.cs) is @benaadams idea
[NoLookupBranchless](https://github.com/gfoidl/Benchmarks/blob/0fe785ac88273e977adf44c1dcd8891651316420/aspnet/Kestrel/CorrelationIdGenerator_PR2579/CorrelationIdGenerator/ChunkWriters/ChunkWriter4.cs) is @benaadams idea in a branchless variant.

I've update the PR to use `NoLookup` even if it minimal slower than `BytesSafeHexUnsafe`, but there's no unsafe anymore.

For me it's interesting that the branchless version is slower than a branching version (seems that the branch-predictor does a good job):
```asm
; version with branches
; ...
       8BFB                 mov      edi, ebx
       C1FF18               sar      edi, 24
       83E70F               and      edi, 15
       83FF0A               cmp      edi, 10
       7C07                 jl       SHORT G_M56191_IG05
       BE57000000           mov      esi, 87
       EB05                 jmp      SHORT G_M56191_IG06

G_M56191_IG05:
       BE30000000           mov      esi, 48

G_M56191_IG06:
       03FE                 add      edi, esi
       400FB6FF             movzx    rdi, dil
       40887811             mov      byte  ptr [rax+17], dil
; ...

; branchless version
; ...
       8BFB                 mov      edi, ebx
       C1FF18               sar      edi, 24
       83E70F               and      edi, 15
       8D77F6               lea      esi, [rdi-10]
       C1FE1F               sar      esi, 31
       83E667               and      esi, 103
       83F657               xor      esi, 87
       03FE                 add      edi, esi
       400FB6FF             movzx    rdi, dil
       40887811             mov      byte  ptr [rax+17], dil
; ...
```

Though I would prefer anyway the version with the ternary and let the JIT emit proper code.

Reply:
I'll try to do the same for `CorrelationIdGenerator` and `StringUtilities`.